pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,774 @@
+	conditions := []crdv1alpha1.PacketCaptureCondition{}
+
+	if status == packetCapturePhaseRunning {
+		conditions = append(conditions, crdv1alpha1.PacketCaptureCondition{
+			Type:               crdv1alpha1.PacketCaptureRunning,
+			Status:             metav1.ConditionStatus(v1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+		})
+	} else if status == packetCapturePhasePending || status == packetCapturePhaseUnknown {
+		conditions = append(conditions, crdv1alpha1.PacketCaptureCondition{
+			Type:               crdv1alpha1.PacketCapturePending,
+			Status:             metav1.ConditionStatus(v1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+		})
+	} else {
+		return nil
+	}
+
+	toUpdate, getErr := c.packetCaptureLister.Get(name)
+	if getErr != nil {
+		return getErr
+	}
+	if retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		newCond := mergeConditions(toUpdate.Status.Conditions, conditions)
+		if conditionSliceEqualsIgnoreLastTransitionTime(toUpdate.Status.Conditions, newCond) {
+			return nil
+		}
+		toUpdate.Status.Conditions = newCond
+		klog.V(2).InfoS("Updating PacketCapture", "name", name, "status", toUpdate.Status)
+		_, updateErr := c.crdClient.CrdV1alpha1().PacketCaptures().UpdateStatus(context.TODO(), toUpdate, metav1.UpdateOptions{})
+		if updateErr != nil && apierrors.IsConflict(updateErr) {
+			var getErr error
+			if toUpdate, getErr = c.crdClient.CrdV1alpha1().PacketCaptures().Get(context.TODO(), name, metav1.GetOptions{}); getErr != nil {
+				return getErr
+			}
+		}
+		return updateErr
+	}); retryErr != nil {
+		return retryErr
+	}
+
+	return nil
+}
+
+func (c *Controller) updatePacketCaptureStatus(name string, num int32, path string, err error) error {
+	toUpdate, getErr := c.packetCaptureLister.Get(name)
+	if getErr != nil {
+		klog.InfoS("Didn't find the original PacketCapture, skip updating status", "name", name)
+		return nil
+	}
+
+	t := metav1.Now()
+	updatedStatus := crdv1alpha1.PacketCaptureStatus{
+		NumberCaptured: num,
+		FilePath:       path,
+	}
+
+	if err != nil {
+		updatedStatus.FilePath = ""
+		if isCaptureCompleted(toUpdate, num) {
+			updatedStatus.Conditions = []crdv1alpha1.PacketCaptureCondition{
+				{
+					Type:               crdv1alpha1.PacketCaptureCompleted,
+					Status:             metav1.ConditionStatus(v1.ConditionTrue),
+					LastTransitionTime: t,
+					Reason:             "Succeed",
+				},
+			}
+		} else {
+			updatedStatus.Conditions = []crdv1alpha1.PacketCaptureCondition{
+				{
+					Type:               crdv1alpha1.PacketCaptureCompleted,
+					Status:             metav1.ConditionStatus(v1.ConditionFalse),
+					LastTransitionTime: t,
+					Reason:             "CaptureFailed",
+					Message:            err.Error(),
+				},
+			}
+		}
+		if toUpdate.Spec.FileServer != nil {
+			updatedStatus.Conditions = append(updatedStatus.Conditions, crdv1alpha1.PacketCaptureCondition{
```

Comment:
is it guaranteed in this case that we attempted an upload and that it failed?

Reply:
Not exactly. My intention was : if there a capture error, and user have set a file server url in the spec , also mark the upload condition failed. even it didn't go that far.

Reply:
It sure seems unnecessary when capture failed. added a `path != ""` to make sure if only happens after the capture finished.